ignition_ros2_control/src/ignition_ros2_control_plugin.cpp

```diff
@@ -0,0 +1,400 @@
+            continue;
+          }
+        default:
+          {
+            ignwarn << "[ignition_ros2_control] Joint [" << jointName <<
+              "(Entity=" << jointEntity << ")] is of unknown type.\n";
+            continue;
+          }
+      }
+      output[jointName] = jointEntity;
+    }
+
+    return output;
+  }
+};
+
+//////////////////////////////////////////////////
+IgnitionROS2ControlPlugin::IgnitionROS2ControlPlugin()
+: dataPtr(std::make_unique<IgnitionROS2ControlPluginPrivate>())
+{
+}
+
+//////////////////////////////////////////////////
+void IgnitionROS2ControlPlugin::Configure(
+  const ignition::gazebo::Entity & _entity,
+  const std::shared_ptr<const sdf::Element> & _sdf,
+  ignition::gazebo::EntityComponentManager & _ecm,
+  ignition::gazebo::EventManager &)
+{
+  // Make sure the controller is attached to a valid model
+  const auto model = ignition::gazebo::Model(_entity);
+  if (!model.Valid(_ecm)) {
+    ignerr << "[Ignition ROS 2 Control] Failed to initialize because [" <<
+      model.Name(_ecm) << "(Entity=" << _entity <<
+      ")] is not a model. Please make sure that"
+      " Ignition ROS 2 Control is attached to a valid model.\n";
+    return;
+  }
+
+  // Get params from SDF
+  std::string paramFileName = _sdf->Get<std::string>("parameters");
+
+  if (paramFileName.empty()) {
+    ignerr << "Ignition ros2 controle found an empty parameters file. " <<
+      "Failed to initialize.";
+    return;
+  }
+
+  std::vector<std::string> arguments = {"--ros-args", "--params-file", paramFileName};
+
+  std::vector<const char *> argv;
+  for (const auto & arg : arguments) {
+    argv.push_back(reinterpret_cast<const char *>(arg.data()));
+  }
+
+  if (!rclcpp::ok()) {
+    rclcpp::init(static_cast<int>(argv.size()), argv.data());
+    this->dataPtr->node = rclcpp::Node::make_shared("ignition_ros2_control");
+  }
+  this->dataPtr->executor_ = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
+  this->dataPtr->executor_->add_node(this->dataPtr->node);
+  auto spin = [this]()
+    {
+      while (rclcpp::ok()) {
+        this->dataPtr->executor_->spin_once();
+      }
+    };
+  this->dataPtr->thread_executor_spin_ = std::thread(spin);
+
+  RCLCPP_ERROR_STREAM(
+    this->dataPtr->node->get_logger(), "[Ignition ROS 2 Control] Setting up controller for [" <<
+      "(Entity=" << _entity << ")].\n");
+
+  ignmsg << "[Ignition ROS 2 Control] Setting up controller for [" <<
+    model.Name(_ecm) << "(Entity=" << _entity << ")].\n";
+
+  // Get list of enabled joints
+  auto enabledJoints = this->dataPtr->GetEnabledJoints(
+    _entity,
+    _ecm);
+
+  // Read urdf from ros parameter server then
+  // setup actuators and mechanism control node.
+  // This call will block if ROS is not properly initialized.
+  std::string urdf_string;
+  std::vector<hardware_interface::HardwareInfo> control_hardware;
+  try {
+    urdf_string = this->dataPtr->getURDF(this->dataPtr->robot_description_);
```

Comment:
`getURDF` is only called here, with `this->dataPtr->robot_description_`. Are there plans to call this function from some other place? Otherwise, we don't need to pass this in as a parameter, because the function already has access to it.